chain/chain/src/chain.rs

```diff
@@ -1077,62 +1090,65 @@ impl Chain {
 
                 Ok(head)
             }
-            Err(e) => match e.kind() {
-                ErrorKind::Orphan => {
-                    let tail_height = self.store.tail()?;
-                    // we only add blocks that couldn't have been gc'ed to the orphan pool.
-                    if block.header().height() >= tail_height {
+            Err(e) => {
+                match e.kind() {
+                    ErrorKind::InvalidSignature => return Err(e),
```

Comment:
Should not block validity errors be handled here?

If now one sends me a valid header with a tampered body during sync, would I not end up saving it as "processed" here, and then not be able to accept it later?

(if it is a real issue, ideally first add a chain test that catches it)